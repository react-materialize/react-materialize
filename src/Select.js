import React, { Component, Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Icon from './Icon';
import idgen from './idgen';
import constants from './constants';

class Select extends Component {
  constructor(props) {
    super(props);

    this.id = props.id || idgen();
    this.handleChange = this.handleChange.bind(this);
    this.state = { value: props.value };
  }

  handleChange(event) {
    const { onChange } = this.props;
    let value = event.target.value;

    if (this.props.multiple) {
      value = this.instance.getSelectedValues();
    }

    if (onChange) onChange(event);

    this.setState({ value });
  }

  componentDidMount() {
    const { options } = this.props;

    if (typeof M !== 'undefined') {
      this.instance = M.FormSelect.init(this._selectRef, options);
    }
  }

  componentDidUpdate() {
    const { options } = this.props;

    if (this.instance) {
      this.instance.destroy();
    }

    this.instance = M.FormSelect.init(this._selectRef, options);
  }

  componentWillUnmount() {
    if (this.instance) {
      this.instance.destroy();
    }
  }

  render() {
    const {
      s,
      m,
      l,
      xl,
      disabled,
      noLayout,
      browserDefault,
      icon,
      label,
      selectClassName,
      success,
      error,
      validate,
      children,
      multiple,
      ...other
    } = this.props;

    const sizes = { s, m, l, xl };

    let responsiveClasses;
    if (!noLayout) {
      responsiveClasses = { col: true };
      constants.SIZES.forEach(size => {
        responsiveClasses[size + sizes[size]] = sizes[size];
      });
    }

    const wrapperClasses = cx('input-field', responsiveClasses);

    const selectProps = {
      type: 'select',
      id: this.id,
      value: this.state.value,
      disabled,
      multiple,
      ...other
    };

    const renderLabel = () =>
      label && (
        <label
          data-success={success}
          data-error={error}
          htmlFor={selectProps.id}
        >
          {label}
        </label>
      );

    const renderIcon = () => icon && <Icon className="prefix">{icon}</Icon>;

    const renderOption = child =>
      cloneElement(child, { key: child.props.value });

    const renderOptions = () => Children.map(children, renderOption);

    return (
      <div className={wrapperClasses}>
        {renderIcon()}
        <select
          value={this.state.value}
          ref={el => {
            this._selectRef = el;
          }}
          onChange={this.handleChange}
          className={cx(
            {
              validate,
              multiple,
              ['browser-default']: browserDefault
            },
            selectClassName
          )}
          {...selectProps}
        >
          {renderOptions()}
        </select>
        {renderLabel()}
      </div>
    );
  }
}

Select.propTypes = {
  /*
   * Use browser default styles
   */
  browserDefault: PropTypes.bool,
  /*
   * Strip away all layout classes such as col and sX
   */
  noLayout: PropTypes.bool,
  /*
   * Responsive size for small size screens (Mobile Devices <= 600px)
   */
  s: PropTypes.number,
  /*
   * Responsive size for middle size screens (Tablet Devices > 600px)
   */
  m: PropTypes.number,
  /*
   * Responsive size for large size screens (Desktop Devices > 992px)
   */
  l: PropTypes.number,
  /**
   * Responsive size for extra large screens (Large Desktop Devices > 1200px)
   */
  xl: PropTypes.number,
  /*
   * disabled input
   */
  disabled: PropTypes.bool,
  /*
   * override id
   * @default idgen()
   */
  id: PropTypes.string,
  /*
   * prefix icon
   */
  icon: PropTypes.string,
  /*
   * label text
   */
  label: PropTypes.string,
  /*
   * Input initial value
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  /*
   * Add validate class to input
   */
  validate: PropTypes.bool,
  /*
   * Custom success message
   */
  success: PropTypes.string,
  /*
   * Custom error message
   */
  error: PropTypes.string,
  /*
   * Additional classes for input
   */
  selectClassName: PropTypes.string,
  /*
   * override type="text"
   */
  type: PropTypes.string,
  /*
   * onChange callback
   */
  onChange: PropTypes.func,
  /*
   * Render a multiple dropdown,
   * use instance.getSelectedValues()
   * to get array of selected values
   */
  multiple: PropTypes.bool,
  children: PropTypes.any,
  /**
   * Options for the select
   * <a target="_blank" href="https://materializecss.com/select.html#options">https://materializecss.com/select.html</a>
   */
  options: PropTypes.shape({
    classes: PropTypes.string,
    /**
     * Options for the dropdown
     * <a target="_blank" href="http://materializecss.com/dropdown.html#options">http://materializecss.com/dropdown.html</a>
     */
    dropdownOptions: PropTypes.shape({
      alignment: PropTypes.oneOf(['left', 'right']),
      autoTrigger: PropTypes.bool,
      constrainWidth: PropTypes.bool,
      container: PropTypes.node,
      coverTrigger: PropTypes.bool,
      closeOnClick: PropTypes.bool,
      hover: PropTypes.bool,
      inDuration: PropTypes.number,
      outDuration: PropTypes.number,
      onOpenStart: PropTypes.func,
      onOpenEnd: PropTypes.func,
      onCloseStart: PropTypes.func,
      onCloseEnd: PropTypes.func
    })
  })
};

Select.defaultProps = {
  options: {
    classes: '',
    dropdownOptions: {
      alignment: 'left',
      autoTrigger: true,
      constrainWidth: true,
      container: null,
      coverTrigger: true,
      closeOnClick: true,
      hover: false,
      inDuration: 150,
      outDuration: 250,
      onOpenStart: null,
      onOpenEnd: null,
      onCloseStart: null,
      onCloseEnd: null
    }
  }
};

export default Select;
