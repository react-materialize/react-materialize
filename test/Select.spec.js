import React from 'react';
import { shallow, mount } from 'enzyme';
import Select from '../src/Select';
import mocker from './helper/new-mocker';

describe('<Select />', () => {
  let wrapper;

  test('handles layout classes', () => {
    wrapper = shallow(<Select s={4} m={6} />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeTruthy();
    expect(wrapper.hasClass('s4')).toBeTruthy();
    expect(wrapper.hasClass('m6')).toBeTruthy();
  });

  test('handles removing layout classes', () => {
    wrapper = shallow(<Select s={4} m={6} noLayout />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeFalsy();
    expect(wrapper.hasClass('s4')).toBeFalsy();
    expect(wrapper.hasClass('m6')).toBeFalsy();
  });

  test('renders browserDefault', () => {
    wrapper = shallow(<Select browserDefault />);
    expect(wrapper.find('select').hasClass('browser-default')).toBeTruthy();
  });

  test('renders label', () => {
    wrapper = shallow(<Select label="Name" />);
    expect(wrapper.find('label').text()).toEqual('Name');
  });

  test('handles validate', () => {
    wrapper = shallow(<Select validate />);
    expect(wrapper.find('select.validate')).toHaveLength(1);
  });

  test('handles selectClassName', () => {
    wrapper = shallow(<Select validate selectClassName={'hello red'} />);
    expect(wrapper.find('select.hello.red.validate')).toHaveLength(1);
  });

  test('calls onChange if provided', () => {
    wrapper = shallow(<Select />);
    wrapper.setState({ value: 'hello' });
    expect(wrapper.find('select').prop('value')).toEqual('hello');
  });

  test('with icon', () => {
    wrapper = shallow(<Select icon="cloud" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders options', () => {
    wrapper = shallow(
      <Select>
        <option value={1}>1</option>
        <option value={2}>2</option>
      </Select>
    );

    expect(wrapper.find('option')).toHaveLength(2);
  });

  test('renders multiple', () => {
    wrapper = shallow(<Select multiple />);
    expect(wrapper.find('select.multiple')).toHaveLength(1);
  });

  test('handles other attributes', () => {
    wrapper = shallow(<Select name="foo" />);
    expect(wrapper.find('select').prop('name')).toBe('foo');
  });

  describe('initialises', () => {
    const selectInitMock = jest.fn();
    const selectInstanceDestroyMock = jest.fn();
    const selectMock = {
      init: (el, options) => {
        selectInitMock(options);
        return {
          destroy: selectInstanceDestroyMock
        };
      }
    };
    const restore = mocker('FormSelect', selectMock);

    beforeEach(() => {
      selectInitMock.mockClear();
      selectInstanceDestroyMock.mockClear();
    });

    afterAll(() => {
      restore();
    });

    test('calls FormSelect', () => {
      mount(<Select />);

      expect(selectInitMock).toHaveBeenCalledTimes(1);
    });

    test('with default options if none are given', () => {
      mount(<Select />);

      expect(selectInitMock).toHaveBeenCalledWith({
        classes: '',
        dropdownOptions: {
          alignment: 'left',
          autoTrigger: true,
          constrainWidth: true,
          container: null,
          coverTrigger: true,
          closeOnClick: true,
          hover: false,
          inDuration: 150,
          outDuration: 250,
          onOpenStart: null,
          onOpenEnd: null,
          onCloseStart: null,
          onCloseEnd: null
        }
      });
    });
  });
});
