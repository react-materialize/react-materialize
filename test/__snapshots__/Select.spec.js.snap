// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> with icon 1`] = `
<div
  className="input-field col"
>
  <Icon
    className="prefix"
  >
    cloud
  </Icon>
  <select
    className=""
    id="mockId"
    onChange={[Function]}
    options={
      Object {
        "classes": "",
        "dropdownOptions": Object {
          "alignment": "left",
          "autoTrigger": true,
          "closeOnClick": true,
          "constrainWidth": true,
          "container": null,
          "coverTrigger": true,
          "hover": false,
          "inDuration": 150,
          "onCloseEnd": null,
          "onCloseStart": null,
          "onOpenEnd": null,
          "onOpenStart": null,
          "outDuration": 250,
        },
      }
    }
    type="select"
  />
</div>
`;
